### PR TITLE
fix: defi page icon for FOXY in Overview tab changed from FOX to FOXY

### PR DESCRIPTION
**Pull request by irohdw**
Added replaceIcon function to src/pages/defi/components/OpportunityCard.tsx. It is called in the AssetIcon tag in line 113.

Added a commented out replaceValue function in line 18 src/pages/defi/components/OpportunityCardList.tsx. It could be used to replace the icon by swapping the FOX asset ID for FOXY.

If the icon for FOX moves to a different URL, then this code will not replace the icon properly.

## Notice

- [ ] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue
https://github.com/shapeshift/web/issues/1540

### CODE REVIEW

**Comment by gomesalexandre** [comment]
@irohdw Is this one now ready for review?

**Comment by irohdw** [comment]
> @irohdw Is this one now ready for review?

Yes it is.

File: src/pages/Defi/components/OpportunityCardList.tsx [hunk 1]
```diff
@@ -1,17 +1,34 @@
...
+
+export const overrideIconAssetId = (asset: Asset): string => {
+  if (!!asset) return iconReplaceMap.get(asset.assetId) || asset.icon
+  else return ''
+}
+
 export const OpportunityCardList = ({ balances }: { balances: UseEarnBalancesReturn }) => {
   const activeOpportunities = balances.opportunities.filter(o => bnOrZero(o.cryptoAmount).gt(0))
+  const foxyIcon = useAppSelector(state => selectAssetById(state, foxyID)).icon
+  iconReplaceMap.set(foxID, foxyIcon)
```
Comment (0xdef1cafe) [inline src/pages/Defi/components/OpportunityCardList.tsx:31]: this is unnecessary, and asset specific logic leaking into a view.

File: src/pages/Defi/components/OpportunityCardList.tsx [hunk 2]
```diff
@@ -1,17 +1,34 @@
...
 
+import { selectAssetById } from '../../../state/slices/assetsSlice/selectors'
+import { useAppSelector } from '../../../state/store'
 import { UseEarnBalancesReturn } from '../hooks/useEarnBalances'
 import { OpportunityCard } from './OpportunityCard'
+
+export const iconReplaceMap = new Map<AssetId, string>()
+
+const foxID = 'eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d'
+const foxyID = 'eip155:1/erc20:0xdc49108ce5c57bc3408c3a5e95f3d864ec386ed3'
```
Comment (0xdef1cafe) [inline src/pages/Defi/components/OpportunityCardList.tsx:21]: this should be scoped to the function they're used in

File: src/pages/Defi/components/OpportunityCardList.tsx [hunk 3]
```diff
@@ -39,7 +56,13 @@ export const OpportunityCardList = ({ balances }: { balances: UseEarnBalancesRet
         gridGap={6}
       >
         {activeOpportunities.map(opportunity => {
-          return <OpportunityCard key={opportunity.assetId} {...opportunity} />
+          return (
+            <OpportunityCard
+              key={opportunity.assetId}
+              {...opportunity}
+              overrideAssetIconID={overrideIconAssetId}
```
Comment (0xdef1cafe) [inline src/pages/Defi/components/OpportunityCardList.tsx:63]: why does this have two different names? `overrideAssetIconID` doesn't make sense, capitalization isn't consistent.

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
comments

**Comment by gomesalexandre** [comment]
@irohdw Can you fill in the missing description fields, including linking https://github.com/shapeshift/web/issues/1540 as an issue? It's not possible to review this without the right context.

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
See comment.

From a high-level, not a fan of having such granular, asset-specific logic so close to components that are not asset-specific and map over many assets.

The idea is we try to have all the data fetching/parsing/selecting and deriving done in a very few contained places, and react to these. That solves (amongst other issues) discrepancies in logic, where you would have to repeat the logic in every component and possibly forget it in others.

That's precisely the case here: in #1889, `DeFi tab` is mentioned as a scope for these changes. That means that with such implementation, you'd have to duplicate the component-colocated logic for the Earn tab as well, else we'd still have the FOX icon in earn:

image

Have a look at `useFoxyBalances`, which returns the foxy opportunities, and `useNormalizeOpportunities` which normalizes FOXy/Yearn/Cosmos staking opportunities
https://github.com/shapeshift/web/blob/fd67442e47da2136ddf7148fa3549c330a1b89e5/src/pages/Defi/hooks/useFoxyBalances.tsx#L96
https://github.com/shapeshift/web/blob/af9483696fa1430802b71146def53c893415a2d7/src/features/defi/helpers/normalizeOpportunity.tsx#L165

Can't we make it so that the opportunity icon is present in the `EarnOpportunityType` (it is currently not, we get it from `Asset`), which will allow us to do that FOX -> FOXy icon transformation in opportunities parsing/normalizing?

**Pull request closed without merging**